### ISSUE TITLE
Move magic strings to constants

### ISSUE THREAD

**Issue by chlupac** [oqtane/oqtane.framework#280]
Specially Permission strings to extra class

**Comment by sbwalker** [comment]
There is a Utilities.Constants class where I have tried to consolidate these items. Most of the permission strings are in the SiteRepository constructor and I plan to introduce a more robust and extensible solution in this area so that new sites can be created based on custom templates.

**Comment by chlupac** [comment]
This i what I am thinking about, Site templates. 
Maybe simplest solution is serialize existing site to Json.

**Comment by sbwalker** [comment]
One of the philosophies behind Oqtane is encapsulation, so as much as possible I am trying to avoid external resources ( DNN relied on a ton of external resources which was very difficult to maintain/support as the platform evolved over time ). What I am envisioning here is a simple interface ( ie. ISiteTemplate ) which has a Name property and a generic method for CreateSite(Site site). Within this method a developer would have full control over the site creation process. They could created any entities they desire ( ie. users, roles, pages, modules, folders, files, jobs, etc.... ). During installation and site creation the UI would provide an option for the user to choose from a list of templates which are available ( based on classes implementing ISiteTemplate ).

**Comment by chlupac** [comment]
yes, its second option which im thinking about, complex site solution including themes modules and site template

**Comment by chlupac** [comment]
I'm working on similar thing now. Thinking about some helper methods and extensions to build site easily and refactoring friendly, without need of magic strings.
Question is how to describe module typeof(what?).Name
```c#
public static void AddModule<T>(this Page page, ....) where T is IModule
{
}
```


```c#
    public interface ISiteTemplate
    {
        void InstallSite();
        string Name { get; }
    }
```

**Comment by sbwalker** [comment]
I have a partial implementation for the site template feature

**Comment by chlupac** [comment]
great
Look at Permission Builder class, it can help with permission strings on site builder

**Comment by sbwalker** [comment]
I think you may be missing some context for the permissions system in Oqtane so I will need to explain. The concept behind the permissions system  ( and this is the same concept that has been used in DNN for the past 20 years ) is that it is extensible and can be applied to any entity. Currently the core framework defines various permissions for Pages, Modules, Folders, and ModuleDefinitions - however it is possible that more permissions may be defined in the future. For example, it may be possible Pages may want to add Add and Delete permissions ( which would not relevant to any other type of entity ). In addition, a custom module may want to define its own permissions. Imagine a Blog module which has a "Comment" permission. All of these scenarios are currently possible in the existing implementation - it provides support for any entity and any custom permission via a generic permission repository/API, helper methods for serializing permission strings and verifying user access, and a permissions grid which is conifigurable via parameters.

**Comment by sbwalker** [comment]
So based on what I described above, an Enum to define the allowable set of permissions is not a viable solution as it would not be extensible and it would not allow for different types of permissions to be associated to an entity.

**Comment by sbwalker** [comment]
If we truly think permission names are a problem ( ie. magic strings ) then the only real solution is to define some constants for the core framework permissions ie. PageViewPermissionName, PageEditPermissionsName, etc... and use those names rather than hardcoded strings. And custom modules could follow a similar approach by declaring constants within a module for permissions ie. BlogCommentPermissionName. This would centralize these values and provide some value I suppose.

**Comment by chlupac** [comment]
It's no problem, my class (Permission builder) internally works with strings and works with your definition of permissions.
I will change it to this conception. It's only helper class for build/parse/evaluate permission strings.

one thing - Constants class starts to be disorganized.

**Comment by sbwalker** [comment]
Have you looked at the UserSecurity class? It already has some methods for dealing with PermissionStrings. And PermissionRepository does as well - it has methods for converting a List<Permission> to a PermissionString. In general for a core concept such as security it is best to limit the number of classes/methods which deal with this concern. This ensures that all paths execute the same exact logic for evaluating permissions. As soon as there are multiple paths it increases the risk of vulnerabilities and increases the maintenance/support burden. As a result I am very hesitant to introduce new permission classes. An alternative approach to deal with the magic strings in site creation is to not introduce a new permissions class, but use the existing permission methods which already exist ( ie. create  a List<Permission> and use the existing methods )

**Comment by chlupac** [comment]
Yes, I have studied it and have come to the conclusion that there is a missing function to build permissionString. So I tried to write it. I can extend the existing class with these functions.

**Comment by sbwalker** [comment]
PermissionStrings are intended to be an internal concept - a convenient way to create an optimized representation of Permission object collections which can be passed between the client and server. In general, developers should not be manually creating PermissionsStrings - they should be creating Permission object collections and then encoding them. There are a few instances in Oqtane where PermissionsStrings are being manually created today - and it was only implemented this way as a shortcut during development - it is not the recommended approach. This is why I was advocating that the better solution may be to simply modify these few instances to use Permission object collections.

**Comment by chlupac** [comment]
Ok.
let's start from beginning. I will add 
```c#
﻿namespace Oqtane.Shared
{
    public class PermissionNames
    {
        public const string View = "View";
        public const string Edit = "Edit";
        public const string Delete = "Delete";
        public const string Utilize = "Utilize";
    }
}
```
and refactor places where are magic strings used with these functions.

**Comment by sbwalker** [comment]
Currently the core framework has the following permissions:

Page
- View
- Edit

Module
- View
- Edit

Folder
- Browse
- View
- Edit

ModuleDefinition
- Utilize

If we want to create constants for permission names we should declare them in a manner which identifies the entity they belong to. This would help prevent a developer from thinking that a permission like Browse is applicable to Page or Module. So I am fine with creating a dedicated constant class - but they should be named to include entity such as PageEdit or subclassed such as Page.Edit

**Comment by chlupac** [comment]
I allready did replacements in #283, please look at it and merge it before merging naming changes

**Comment by chlupac** [comment]
EntityNames constants will follow

**Comment by sbwalker** [comment]
I cannot merge #283 as the permission name constants which were added do not indicate which entity they belong to. This is what I was describing in my comment above. Permission name constants need to be in the form of EntityNamePermissionName or EntityName.PermissionName.

**Comment by sbwalker** [comment]
Also "Delete" is not currently a permission name for any entity. And Folder Browse is missing.

**Comment by chlupac** [comment]
Browse is added in second commit
Which permissionname put here ???
```c#
        // GET api/<controller>/5
        [HttpGet("{id}")]
        public Setting Get(int id)
        {
            Setting setting = _settings.GetSetting(id);
            if (IsAuthorized(setting.EntityName, setting.EntityId, PermissionNames.View))
            {
                return setting;
            }
            else
            {
                _logger.Log(LogLevel.Error, this, LogFunction.Read, "User Not Authorized To Access Setting {Setting}", setting);
                HttpContext.Response.StatusCode = 401;
                return null;
            }
        }
```

**Comment by chlupac** [comment]
- complicated
- redundant
- do not force developer to use right permission on right place

```c#
    public class PermissionNames
    {
        public const string Browse = "Browse";
        public const string View = "View";
        public const string Edit = "Edit";
        public const string Utilize = "Utilize";

        public const string FolderBrowse = "Browse";
        public const string FolderView = "View";
        public const string FolderEdit = "Edit";

        public const string PageView = "View";
        public const string PageEdit = "Edit";
        
        public const string ModuleView = "View";
        public const string ModuleEdit = "Edit";
        
        public const string ModuleDefinitionUtilize = "Utilize";
    }
```

**Comment by sbwalker** [comment]
I was suggesting:

    public class PermissionNames
    {
        public const string FolderBrowse = "Browse";
        public const string FolderView = "View";
        public const string FolderEdit = "Edit";

        public const string PageView = "View";
        public const string PageEdit = "Edit";
        
        public const string ModuleView = "View";
        public const string ModuleEdit = "Edit";
        
        public const string ModuleDefinitionUtilize = "Utilize";
    }

As this helps a developer identify which permissions belong to which entities. You are correct that it does not enforce it at a code level, but as an example if you were dealing with Pages then it would not make sense to reference the ModuleDefinitionUtilize permission. Whereas if permission names were defined simply as Utilize there is no enforcement but also no hint that this permission name is not appropriate for any entity except for ModuleDefinitions.

**Comment by chlupac** [comment]
And what about places where you do not know correct entity ?

**Comment by sbwalker** [comment]
This is where refactoring in this area gets tricky, as there are concepts in Oqtane which are cross-cutting and do not naturally map to a specific domain/entity.

Settings is a general service similar to Permissions - meaning that it is not specific to an Entity, it handles any Entity. So in this case the code would need to be refactored to check the permission name which is applicable to the entity. So if the EntityName is Module then the code should pass the "ModuleView" permission, and if Page then the code should pass "PageView" permission, etc... There will also be cases where the EntityName does not have an associated permissions aspect  - in which case the private IsAuthorized() method would need to have some custom logic like it currently does for the User entity - or it should not do a security check and just return true and rely on the caller to enforce any security.

**Comment by chlupac** [comment]
Pls, merge this as first step of refactoring, because I have identified lot of places, where are magic strings, then merge naming refactoring, and I will continue.

**Comment by sbwalker** [comment]
Thinking through the scenario for custom modules. Lets say you have a Blog module which contains a Blog entity and you want to use the generic Settings service. You can pass "Blog" as the EntityName from your module code - this would work fine. But in order to enforce security on the Blog Settings, the framework currently makes an assumption that "View" permissions map to a Get Settings request. What if the Blog module decided to use "BlogRead" as its permission name for viewing blogs. This would obviously not currently work as the Settings API has no way to know that the Blog module uses its own custom permission name. So in this case the IsAuthorized() method could ignore the security check and return the Setting and rely on the caller to do its own security check. 

However this would be a potential security vulnerability as the Setting API could return details which are not appropriate ( and security enforcement must always happen on the server - not on the client ).

One potential solution is to allow the caller to pass an explicit "PermissionName" parameter to the Settings API Get() method so that the framework knows which permission name it should validate. This would seem to be the best solution as it makes no assumptions and ensures security enforcement happens on the server.

Another option is to still support a generic "View" and "Edit" permission name - but the system would then be relying on convention and assumptions which is generally not a good long term solution.

**Comment by sbwalker** [comment]
The problem with merging this PR now is that the full impact/ramifications have not been fully explored yet - which means there is a very high probability of disruption to the integrity of the system - which is an even higher concern because this is the core security system of the framework.

**Comment by sbwalker** [comment]
Commenting on your statement above:

"We should validate used permissions and throw exception when wrong permission is used.
You are mainly using these constants in two places:    PermissionString creation - Entity can validate string against set of allowed permissions"

This is not possible. You are not considering the use case that this service is used by custom modules which can define their own permissions ie. BlogComment permission, etc... The list of permissions is not static - it is dynamic as every module has the ability to define their own permissions.

**Comment by chlupac** [comment]
of course, and it's only about definition of allowed permission on module, simple list like you use on permission grid, yes i'm considering this.

**Comment by chlupac** [comment]
public virtual string[] UsedPermissions => new[]{PermissionNames.Edit, BlogPermission.Comment}

**Comment by sbwalker** [comment]
Lets back this up for a moment. We have an existing security system which is functional. The goal of this PR was to simply replace some magic strings with constants ( ie. some minor code cleanup ). In the process it became clear that the extensibility aspects of the permission and settings areas were not well understood ( which makes sense because the only use cases are currently implemented in the core framework - which does not represent the full picture ). It also exposed a design flaw in the Settings area where it makes assumptions about permissions and is not fully extensibility. But so far there is nothing else which has been identified which would require deep refactoring in order to accomplish the original goal. Correct?

**Comment by chlupac** [comment]
100% agree

**Comment by chlupac** [comment]
Original goal is code cleanup, it's fulfilled with this PR. Everything else is extension.

**Comment by sbwalker** [comment]
So the simplest and least disruptive path forward is to replace the existing magic strings ie. View, Edit, etc... with associated constants. This preserves the integrity of the system. In a subsequent step we could explore refactoring the constants to be related to specific entities ( ie. ModuleView, PageEdit, etc... ) and deal with the design flaw in Settings.

**Comment by sbwalker** [comment]
so in reviewing #283 I am still seeing a Delete permission? Or am I missing something?

**Comment by chlupac** [comment]
done in last commit

**Comment by sbwalker** [comment]
I just want to say thank you for the productive discussion in this area... and for your patience. If we had not gone down the proverbial rabbit hole, it would not have revealed the design flaw or allowed for the sharing of knowledge related to the expectations or extensibility of permissions in the system. I appreciate your help very much.

**Comment by chlupac** [comment]
Next magic string set - entity names
any specifications to this area ?

**Comment by chlupac** [comment]
and here is still 1694 warnings in code according to R#

**Comment by sbwalker** [comment]
Entity names should be more straightforward ;) It might make sense to resolve the few conflicts in #284 before you start on entity names so that I can merge it.

**Comment by chlupac** [comment]
ok

**Comment by chlupac** [comment]
What do you say about this?
```c#
    public class PermissionString
    {
        public string PermissionName { get; set; }
        public string Permissions { get; set; }

        public PermissionString(string permissionName, string permissions = null)
        {
            PermissionName = permissionName;
            Permissions = permissions;
        }

        public PermissionString GrantTo(params string[] roles)
        {
            if (!roles.Any()) return this;
            var list = string.IsNullOrEmpty(Permissions) ? roles : Permissions.Split(";").Concat(roles);
            Permissions = string.Join(";", list.Distinct());
            return this;
        }

        public PermissionString DenyTo(params string[] roles)
        {
            if (!roles.Any()) return this;
            var list = string.IsNullOrEmpty(Permissions)
                ? roles.Select(r => "!" + r)
                : Permissions.Split(";")
                    .Concat(roles.Select(r => "!" + r));
            
            Permissions = string.Join(";", list.Distinct());
            return this;
        }
```
usage:
```c#
var ps = new PermissionString(PermissionNames.Edit).GrantTo(Constants.AdminRole, Constants.HostRole).DenyTo(Constants.AllUsersRole);
```

**Comment by sbwalker** [comment]
A PermissionString is essentially an ACL that is used internally by the framework for optimized authorization. PermissionStrings are serialized in the form of:

"{permissionname:!rolename1;![userid1];rolename2;rolename3;[userid2];[userid3]}" 

Where "!" designates Deny permissions and deny permissions should be stored at the beginning of the string and grant permissions at the end ( for permormance - as this exact concept was used in DNN based and was developed based on performance testing various approaches ).

So I am not against having a constructor for the class - although its not really necessary as C# allows us to initialize members of a class without a constructor. And as I mentioned earlier in this thread, developers should NOT be manually creating PermissionString objects as they are supposed to be an internal concept only. I am not using them in the refactoring of SiteTemplates.

Based on what I described above the GrantTo and DenyTo methods do not make sense as they do not handle users - only roles. And they duplicate internal logic related to how permissions strings are encoded which is a bad practice - such logic should exist only in a single location. 

My recommendation here is that we should not make any changes to PermissionStrings as it has more disadvantages than advantages - and its use be developers should not be encouraged,

**Comment by chlupac** [comment]
ok, forgot this
